app/models/assistant/export.rb

```diff
@@ -0,0 +1,53 @@
+module Assistant::Export
+  extend ActiveSupport::Concern
+
+  DEFAULT_EXPORT_ONLY = %i[
+    name
+    description
+    instructions
+    tools
+    external_id
+    language_model_api_name
+  ]
+
+  DEFAULT_ASSISTANT_FILE = "assistants.yml"
+
+  def attributes
+    super.merge("language_model_api_name" => language_model_api_name)
+  end
+
+  # Unsure why this needs to re-defined, but the original ActiveModel::Serialization
+  # implementation is ignoring the #attributes method above.
+  def attribute_names_for_serialization
+    attributes.keys
+  end
+
+  class_methods do
+    def export_to_file(path: Rails.root.join(DEFAULT_ASSISTANT_FILE), assistants:, only: DEFAULT_EXPORT_ONLY)
+      path = path.to_s
+      storage = {
+        "assistants" => assistants.as_json(only:).map(&:compact)
+      }
+      if path.ends_with?(".json")
+        File.write(path, storage.to_json)
+      else
+        File.write(path, storage.to_yaml)
+      end
+    end
+
+    def import_from_file(path: Rails.root.join(DEFAULT_ASSISTANT_FILE), users: User.all)
+      users = Array.wrap(users)
+
+      storage = YAML.load_file(path)
+      assistants = storage["assistants"]
+      assistants.each do |assistant|
+        assistant = assistant.with_indifferent_access
+        users.each do |user|
+          asst = user.assistants.find_or_create_by(name: assistant["name"])
+          asst.assign_attributes(assistant.except("name"))
+          asst.save!
+        end
+      end
+    end
```

Comment:
I tested this locally and I'm thinking through a common workflow:

Say OpenAI comes out with a new variation of gpt-4o tomorrow called gpt-4.5o. We'll tee up a PR that updates the `models.yml` with this new model. It'll have a distinct `api_name` and we'd name it `GPT 4.5o (latest)` and `best: true`. Then we would update the existing gpt-4o model in our to change `best: false` and remove `(latest)` from it's name or maybe changing that parenthetical to the date of it's original date.

All of this seems doable since `language_models/export.rb` does find_or_initialize_by(api_name). Good call on that.

Then, I'd normally write the migrations to remove GPT-4o from everyone's assistant and instead add a new GPT-4.5o as their assistant. But this new assistants.yml workflow doesn't support that.

Should it? Maybe not. If we were to merge this in as-is, we'd (a) change the `GPT-4o` entry within it to `GPT-4.5o` and we could write a migration to remove the existing one for everyone.

What if we wanted to do a more minor edit such as changing `GPT-4o` to `GPT 4o` and changing `Meta Llama` to `Meta Llama 70b`. If we made those changes within assistants.yml then it would create new duplicate ones.

**I think we should consider this update logic to (1) grab the language_model record based on `language_model_api_name` and then (2) for each user do a `find_or_initialize_by(language_model_id: )`. That would permit us to rename assistants for people.**

So that would enable our YML files to be useful for adding models & assistants, changing details of models & assistants, but not removing either. We use migrations to remove things. That feels right. Do you agree?

Reply:
> I think we should consider this update logic to (1) grab the language_model record based on language_model_api_name and then (2) for each user do a find_or_initialize_by(language_model_id: ). That would permit us to rename assistants for people.

So I think my own uses of Assistants will have many that share the same LLM (e.g. `gpt-4o` latest) but have different system prompts. I think this might be called an Agent to some people.

Therefore neither find/update by `name` (original PR) nor `language_model_api_name` is appropriate.

Couple ideas:

1. We introduce a slug for each Assistant that is permanent, even if the name is changed. E.g. "GPT 4o" could have slug `gpt-4o`. Now, many Assistants in the system would have duplicate `gpt-4o` but we'd only allow one per user. We could then use `User.assistants.find_or_initialize_by(slug:)` for the update.
2. We never update, only create Assistants for new users.

Reply:
Oh, good point. I think what you’re doing with multiple assistants is important to support.

I like the idea of a slug as the way to resolve it. That’s a good idea. I think it’s worth us having a small set of default assistants for users which is basically the “best” assistant for each service, so the slug could be something like “openai-default” or something like that, which we can then continually update as these LLMs evolve.

Reply:
Ok, PR now has `Assistant#slug` support; including migration that tries hard to map any existing Assistant's to the "best" assistants' slugs in assistants.yml; and generates unique slugs for all other `Assistant` based on `name`.